com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/JarSignatureVerifier.java

```diff
@@ -0,0 +1,175 @@
+        }
+    }
+
+    private void verifyJarEntries(List<X509Certificate> targetCertificates, List<JarEntry> jarEntries, String certificateCN) {
+        List<String> filesWithoutCertificates = new ArrayList<>();
+        for (JarEntry jarEntry : jarEntries) {
+            Certificate[] entryCertificates = jarEntry.getCertificates();
+            if (entryCertificates == null || entryCertificates.length == 0) {
+                verifyIsMetaInformation(jarEntry, filesWithoutCertificates);
+                continue;
+            }
+            verifyJarEntryIsSigned(targetCertificates, entryCertificates, certificateCN);
+        }
+        if (!filesWithoutCertificates.isEmpty()) {
+            throw new SLException(MessageFormat.format(Messages.ARCHIVE_CONTAINS_UNSIGNED_FILES, filesWithoutCertificates));
+        }
+    }
+
+    private void verifyIsMetaInformation(JarEntry jarEntry, List<String> filesWithoutCertificates) {
+        if (!jarEntry.getName()
+                     .startsWith(META_INF)) {
+            filesWithoutCertificates.add(jarEntry.getName());
+        }
+    }
+
+    private void verifyJarEntryIsSigned(List<X509Certificate> targetCertificates, Certificate[] entryCertificates, String certificateCN) {
+        List<X509Certificate> entryCertificatesList = toX509Certificates(entryCertificates);
+        if (!isValidCertificateChain(targetCertificates, entryCertificatesList, certificateCN)) {
+            throw new SLException(MessageFormat.format(Messages.THE_ARCHIVE_IS_NOT_SIGNED_BY_TRUSTED_CERTIFICATE_AUTHORITY,
+                                                       getTrustedCertificatesNames(targetCertificates)));
+        }
+    }
+
+    private List<X509Certificate> toX509Certificates(Certificate[] certificates) {
+        return Arrays.stream(certificates)
+                     .map(certificate -> (X509Certificate) certificate)
+                     .collect(Collectors.toList());
+    }
+
+    private boolean isValidCertificateChain(List<X509Certificate> targetCertificates, List<X509Certificate> candidateCertificateChain,
+                                            String certificateCN) {
+        boolean certificateCNMatches = false;
+        boolean targetCertificatesMatchEntryCertificate = false;
+        for (X509Certificate certificate : candidateCertificateChain) {
+            checkValidityOfCertificate(certificate);
+            if (!certificateCNMatches) {
+                certificateCNMatches = certificateCNMatches(certificate, certificateCN);
+            }
+            if (!targetCertificatesMatchEntryCertificate) {
+                targetCertificatesMatchEntryCertificate = doTargetCertificatesMatchEntryCertificate(targetCertificates, certificate);
+            }
+        }
+        if (!certificateCNMatches) {
+            throw new SLException(MessageFormat.format(Messages.NONE_OF_YOUR_CERTIFICATES_CNS_MATCH, certificateCN));
+        }
+        return targetCertificatesMatchEntryCertificate;
```

Comment:
It's inconsistent to throw an exception in case the CN doesn't match, and return false in case the certificates don't.